source/server/config_validation/server.cc

```diff
@@ -104,5 +107,31 @@ void ValidationInstance::shutdown() {
   thread_local_.shutdownThread();
 }
 
+void ValidationInstance::initializeStatsTags(const envoy::api::v2::Bootstrap& bootstrap) {
+  // Ensure no tag names are repeated.
+  std::unordered_set<std::string> names;
+  auto add_tag = [&names, this](const std::string& name, const std::string& regex) {
+    if (!names.emplace(name).second) {
+      throw EnvoyException(fmt::format("Tag name '{}' specified twice.", name));
+    }
+
+    tag_extractors_.emplace_back(Stats::TagExtractorImpl::createTagExtractor(name, regex));
+  };
+
+  // Add defaults.
+  if (!bootstrap.stats_config().has_use_all_default_tags() ||
+      bootstrap.stats_config().use_all_default_tags().value()) {
+    for (const std::pair<std::string, std::string>& default_tag :
+         Config::TagNames::get().regex_map_) {
+      add_tag(default_tag.first, default_tag.second);
+    }
+  }
+
+  // Add custom tags.
+  for (const envoy::api::v2::TagSpecifier& tag_specifier : bootstrap.stats_config().stats_tags()) {
+    add_tag(tag_specifier.tag_name(), tag_specifier.regex());
+  }
```

Comment:
Do we need this duplication in the config validation implementation? I'm hoping we can avoid rewriting logic like this where possible. Can it be refactored if it is needed?